src/client/datascience/notebook/notebookControllerManager.ts

```diff
@@ -52,10 +50,11 @@ export class NotebookControllerManager implements INotebookControllerManager, IE
     private controllersPromise?: Promise<void>;
 
     // Listing of the controllers that we have registered
-    private registeredControllers: VSCodeNotebookController[] = [];
+    private registeredControllers = new Map<string, VSCodeNotebookController>();
```

Comment:
Changed to a Map, this way we can discover Non-Python kernels in parallel (as we get new items we just push it into this map & easy to check if an item exists, by id).

Same list is used to re-fetch controllers when Pyton extension is installed (re-run controller discovery without having to worry about duplicate controllers). Easy fix.

Reply:
This ensures we don't need to reload VS Code (at least for kernel discovery, partly fixes https://github.com/microsoft/vscode-jupyter/issues/6353).
I think there are other areas that might require similar fixes. I fixed this as it was super simple